pkg/istio/onboarding/controller_test.go

```diff
@@ -0,0 +1,401 @@
+		return nil, nil
+	}
+
+	clusterRbacConfig, ok := config.Spec.(*v1alpha1.RbacConfig)
+	if !ok {
+		log.Panicln("cannot cast to rbac config")
+	}
+	return config, clusterRbacConfig
+}
+
+func newFakeController(services []*v1.Service, fake bool) *Controller {
+	c := &Controller{}
+	configDescriptor := model.ConfigDescriptor{
+		model.ClusterRbacConfig,
+	}
+
+	store := memory.Make(configDescriptor)
+	if fake {
+		store = &fakeConfigStore{store}
+	}
+	c.store = memory.NewController(store)
+
+	source := fcache.NewFakeControllerSource()
+	for _, service := range services {
+		source.Add(service)
+	}
+	fakeIndexInformer := cache.NewSharedIndexInformer(source, &v1.Service{}, 0, nil)
+	stopChan := make(chan struct{})
+	go fakeIndexInformer.Run(stopChan)
+
+	if !cache.WaitForCacheSync(stopChan, fakeIndexInformer.HasSynced) {
+		log.Panicln("timed out waiting for cache to sync")
+	}
+	c.serviceIndexInformer = fakeIndexInformer
+	c.dnsSuffix = dnsSuffix
+
+	return c
+}
+
+func TestNewController(t *testing.T) {
+	configDescriptor := model.ConfigDescriptor{
+		model.ClusterRbacConfig,
+	}
+
+	source := fcache.NewFakeControllerSource()
+	fakeIndexInformer := cache.NewSharedIndexInformer(source, &v1.Service{}, 0, nil)
+	configStore := memory.Make(configDescriptor)
+	configStoreCache := memory.NewController(configStore)
+
+	c := NewController(configStoreCache, dnsSuffix, fakeIndexInformer)
+	assert.Equal(t, dnsSuffix, c.dnsSuffix, "dns suffix should be equal")
+	assert.Equal(t, fakeIndexInformer, c.serviceIndexInformer, "service index informer pointer should be equal")
+	assert.Equal(t, configStoreCache, c.store, "config store cache pointer should be equal")
+}
+
+func TestProcessEvent(t *testing.T) {
+	c := &Controller{queue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())}
+
+	tests := []struct {
+		name             string
+		inputObject      interface{}
+		expectedQueueLen int
+		expectedKeyName  string
+	}{
+		{
+			name:             "test processing a service object",
+			inputObject:      onboardedService,
+			expectedQueueLen: 1,
+			expectedKeyName:  "test-namespace/onboarded-service",
+		},
+		{
+			name:             "test processing a non service object",
+			inputObject:      nil,
+			expectedQueueLen: 0,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c.processEvent(cache.MetaNamespaceKeyFunc, onboardedService)
+			assert.Equal(t, tt.expectedQueueLen, c.queue.Len(), "expected queue length should be equal")
+
+			if tt.expectedQueueLen > 0 {
+				item, shutdown := c.queue.Get()
+				assert.Equal(t, tt.expectedKeyName, item, "expected key name should be equal")
+				assert.Equal(t, false, shutdown, "shutdown should be false")
+			}
+		})
+	}
+
+	c.queue.ShutDown()
+}
+
+func TestAddService(t *testing.T) {
+	c := &Controller{}
+
+	tests := []struct {
+		name              string
+		inputServices     []string
+		clusterRbacConfig *v1alpha1.RbacConfig
+		expectedArray     []string
+	}{
+		{
```

Comment:
could we add a nil `clusterRbacConfig:` test case?

Reply:
also, an empty `inputServices` list?

Reply:
Sure, added the tests.